app.go

```diff
@@ -261,22 +263,69 @@ func (t stopTimeoutOption) String() string {
 	return fmt.Sprintf("fx.StopTimeout(%v)", time.Duration(t))
 }
 
+// WithLogger will stay private until we export fxlog.Logger once
+// we finalize the API.
+func WithLogger(l fxlog.Logger) Option {
+	return withLoggerOption{l}
+}
+
+type withLoggerOption struct{ l fxlog.Logger }
+
+func (l withLoggerOption) apply(app *App) {
+	app.log = l.l
+}
+
+func (l withLoggerOption) String() string {
+	return fmt.Sprintf("fx.WithLogger(%v)", l.l)
+}
+
 // Printer is the interface required by Fx's logging backend. It's implemented
 // by most loggers, including the one bundled with the standard library.
+//
+// Note, this will be deprecate with next release and you will need to implement
+// fxlog.Logger interface instead.
 type Printer interface {
 	Printf(string, ...interface{})
 }
 
 // Logger redirects the application's log output to the provided printer.
+//
+// Note, this will be deprecate with next release and you will need to use
+// WithLogger instead.
 func Logger(p Printer) Option {
 	return loggerOption{p}
 }
 
 type loggerOption struct{ p Printer }
 
+type printerWrapper struct {
+	p Printer
+}
+
+// newPrinter returns an implementation of zapcore.WriteSyncer
+// used to support Logger option which implements Printer
+// interface.
+func newPrinter(p Printer) zapcore.WriteSyncer {
+	return &printerWrapper{
+		p: p,
+	}
+}
+
+func (p *printerWrapper) Write(b []byte) (n int, err error) {
+	p.p.Printf(string(b))
+
+	return len(b), nil
+}
+
+func (p *printerWrapper) Sync() error {
+	return nil
+}
+
 func (l loggerOption) apply(app *App) {
-	app.logger = &fxlog.Logger{Printer: l.p}
-	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.logger)}
+	np := newPrinter(l.p)
+
+	app.log = fxlog.DefaultLogger(zapcore.Lock(np)) // assuming np isnt thread-safe.
```

Comment:
this may not be necessary. we don't do any concurrent logging in fx, and it may not be unreasonable to ask the caller to make fx.Printer thread safe. before this change, fx.Printer was expected to be a [log.Logger](https://golang.org/pkg/log/#Logger) which is already thread safe.